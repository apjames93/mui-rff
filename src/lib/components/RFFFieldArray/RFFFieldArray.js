import React, { cloneElement } from 'react';
import PropTypes from 'prop-types';
import { FieldArray } from 'react-final-form-arrays';

/**
 * RFFFieldArray
*/

export const RFFFieldArray = ({
  children,
  fieldArrayName,
}) => (
  <FieldArray
    name={fieldArrayName}
  >
    {({ fields }) => (
      <>
        {fields.map((name, index) => (
          <div key={index} style={{ width: '100%' }}>
            {Array.isArray(children) && children.map((child, i) => cloneElement(
              child,
              {
                fields,
                key: i,
                name: child.props.name ? `${name}[${child.props.name}]` : name,
                index,
              },
            ))}
            {!Array.isArray(children) && cloneElement(
              children,
              {
                fields,
                name: children.props.name ? `${name}[${children.props.name}]` : name,
              },
            )}
          </div>
        ))}
      </>
    )}
  </FieldArray>
);

export default RFFFieldArray;

RFFFieldArray.propTypes = {
  children: PropTypes.oneOfType(
    [PropTypes.arrayOf(PropTypes.element), PropTypes.element],
  ).isRequired,
  fieldArrayName: PropTypes.string.isRequired,
};

RFFFieldArray.defaultProps = {};
