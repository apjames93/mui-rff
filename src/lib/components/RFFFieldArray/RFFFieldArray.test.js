import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import arrayMutators from 'final-form-arrays';
import {
  Form, MuiInput, MuiRadio, MuiCheckbox,
} from '../../index';
import RFFFieldArray from './RFFFieldArray';

function setup() {
  const props = { fieldArrayName: 'array' };
  const comp = mount(
    <Form
      onSubmit={() => true}
      mutators={{ ...arrayMutators }}
      initialValues={{
        array: [
          {
            name: 'wooo',
            state: '',
            gender: '',
            canCall: false,
          },
        ],
      }}
    >
      <RFFFieldArray {...props}>
        <MuiInput name="name" label="name" />
        <MuiRadio name="gender" value="male" color="default" label="male" />
        <MuiCheckbox name="canCall" />
      </RFFFieldArray>
    </Form>,
  );
  return { comp, props };
}

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

describe('<RFFFieldArray />', () => {
  it('renders RFFFieldArray', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(
      <Form onSubmit={() => true} mutators={{ ...arrayMutators }}>
        <RFFFieldArray {...props} />
      </Form>,
    );
    expect(tree).toMatchSnapshot();
  });
});
