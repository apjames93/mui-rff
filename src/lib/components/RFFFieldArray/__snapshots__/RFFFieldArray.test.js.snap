// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RFFFieldArray /> snapshot 1`] = `
<form
  onSubmit={[Function]}
/>
`;
